<div class="output_documentation_box">

    <h3>Output Documentation</h3>

    <p>
        Result of the analysis of differentially expressed genes.
        Quantification of the differentially expressed genes was done with Salmon, while
        the differential expression analysis was done with DESeq2.
        Only significant differentially expressed genes (adjusted p-value < 0.10)  are shown.
        The experiment was done between the annotated control samples and one given condition (see workflow
        annotation).
    </p>
    <button class="btn btn-secondary btn-sm" type="button" data-bs-toggle="collapse" data-bs-target="#collapseDiv"
            aria-expanded="false" aria-controls="collapseDiv">
        Show details
    </button>

    <div class="collapse" id="collapseDiv">
        <h5 class="pt-5">Column descriptions:</h5>
        <ol class="list-group pb-5">
            <li class="list-group-item"><b>subject: </b>Gene ID (Ensembl).</li>
            <li class="list-group-item"><b>baseMean: </b>Normalized expression (TPM) mean of all samples in the dataset.</li>
            <li class="list-group-item"><b>log2FoldChange: </b>Estimation: Log-2 Fold Change (Control vs. Condition).</li>
            <li class="list-group-item"><b>lfcSE: </b>Standard error for the log-fold-change estimation.</li>
            <li class="list-group-item"><b>pvalue: </b>Computed p-value (Wald test) for measuring this gene as differentially expressed.</li>
            <li class="list-group-item"><b>padj: </b>Adjusted p-value (Benjamini Hochberg) for measuring this gene as differentially expressed.</li>
            <li class="list-group-item"><b>symbol: </b>Official gene symbol (Entrez)</li>
            <li class="list-group-item"><b>entrez: </b>Gene-ID (Entrez)</li>
        </ol>

        <h5>Method:</h5>
        <p>
            <a href="https://combine-lab.github.io/salmon/">Salmon</a> is a tool for quantifying the expression of transcripts using RNA-seq data.
            Salmon uses new algorithms (specifically, coupling the concept of quasi-mapping with a two-phase
            inference procedure) to provide accurate expression estimates very quickly (i.e. wicked-fast) and while
            using little memory.
            Salmon performs its inference using an expressive and realistic model of RNA-seq data that takes
            into account experimental attributes and biases commonly observed in real RNA-seq data.
        </p>
        <p>
            <a href="https://genomebiology.biomedcentral.com/articles/10.1186/s13059-014-0550-8">DESeq2</a> estimates
            variance-mean dependences in count data from high-throughput sequencing assays and
            tests for differential expression based on a model using the negative binomial distribution.
        </p>
    </div>

</div>
