<div class="output_documentation_box">

    <h3>Output Documentation</h3>

    <p>
        Summary of rMATS' result regarding the detection of differential alternative splicing events from RNA-Seq data.
        <br><br>
        Each alternative splicing event type has a corresponding set of output files:
        [SE (skipped exon), MXE (mutually exclusive exons), A3SS (alternative 3' splice site),
        A5SS (alternative 5' splice site), RI (retained intron)].
        <br>
        Additionally, [JC]-files include only reads that span junctions defined by rmats (Junction Counts), while
        [JCEC]-files include both reads that span junctions defined by rmats (Junction Counts) and reads that do
        not cross an exon boundary (Exon Counts).
        <br><br>
        <b>Attention: </b>Results have been prefiltered to only include significant events (p-value <0.05 and FDR < 0.10).
        If the results are more than 1000 lines, only the first 1000 lines are shown.
    </p>
    <button class="btn btn-secondary btn-sm" type="button" data-bs-toggle="collapse" data-bs-target="#collapseDiv"
            aria-expanded="false" aria-controls="collapseDiv">
        Show details
    </button>

    <div class="collapse" id="collapseDiv">
        <h5 class="pt-5">Column descriptions:</h5>
        <p>
            For event event-specific columns please visit the rMATS-
            <a href="https://github.com/Xinglab/rmats-turbo/blob/v4.1.2/README.md">ReadMe.</a>
        </p>
        <ol class="list-group pb-5">
            <li class="list-group-item"><b>ID: </b>rMATS event id</li>
            <li class="list-group-item"><b>GeneID: </b>Gene id</li>
            <li class="list-group-item"><b>geneSymbol: </b>Gene name</li>
            <li class="list-group-item"><b>chr: </b>Chromosome</li>
            <li class="list-group-item"><b>strand: </b>Strand of the gene</li>
            <li class="list-group-item"><b>IJC_SAMPLE_1: </b>Inclusion counts for sample 1. Replicates are comma separated</li>
            <li class="list-group-item"><b>SJC_SAMPLE_1: </b>Skipping counts for sample 1. Replicates are comma separated</li>
            <li class="list-group-item"><b>IJC_SAMPLE_2: </b>Inclusion counts for sample 2. Replicates are comma separated</li>
            <li class="list-group-item"><b>SJC_SAMPLE_2: </b>Skipping counts for sample 2. Replicates are comma separated</li>
            <li class="list-group-item"><b>IncFormLen: </b>Length of inclusion form, used for normalization</li>
            <li class="list-group-item"><b>SkipFormLen: </b>Length of skipping form, used for normalization</li>
            <li class="list-group-item"><b>PValue: </b>Significance of splicing difference between the two sample groups. (Only available if the statistical model is on)</li>
            <li class="list-group-item"><b>FDR: </b>False Discovery Rate calculated from p-value. (Only available if statistical model is on)</li>
            <li class="list-group-item"><b>IncLevel1: </b>Inclusion level for sample 1. Replicates are comma separated. Calculated from normalized counts</li>
            <li class="list-group-item"><b>IncLevel2: </b>Inclusion level for sample 2. Replicates are comma separated. Calculated from normalized counts</li>
            <li class="list-group-item"><b>IncLevelDifference: </b>average(IncLevel1) - average(IncLevel2)</li>
        </ol>

        <p>
            <b>Summary file: </b>
            Brief summary of all AS event types.
            Includes the total event counts and significant event counts.
            By default, events are counted as significant if FDR <= 0.05.
        </p>

        <h5>Method:</h5>
        <p>
            <a href="https://rnaseq-mats.sourceforge.net/">rMATS</a>: is a computational tool to detect
            differential alternative splicing events from RNA-Seq data. The statistical model of MATS calculates
            the P-value and false discovery rate that the difference in the isoform ratio of a gene between two
            conditions exceeds a given user-defined threshold.
            From the RNA-Seq data, MATS can automatically detect and analyze alternative splicing events corresponding
            to all major types of alternative splicing patterns.
            MATS handles replicate RNA-Seq data from both paired and unpaired study design.
        </p>
    </div>

</div>
